torch/distributed/_shard/sharded_tensor/_ops/chunk.py

```diff
@@ -1,35 +1,235 @@
+        idx_metadata = meta_ranks.index(rank)
+        index = _generate_chunk_indices(
+            chunks,
+            shards_metadata,
+            chunk_size,
+            world_size,
+            idx_metadata,
+            placement_ranks,
+            dim,
         )
-    return results
+        output_split_sizes[rank] = len(index[cur_rank])
+
+    local_tensor = local_tensor.index_select(
+        0, torch.tensor(indices_flatten, device=local_tensor.device)
+    )
+    local_tensor_size = local_tensor.size()
+    gathered_input_size = [sum(output_split_sizes)] + list(local_tensor_size[1:])
+    gathered_input = torch.empty(gathered_input_size, device=local_tensor.device)
+    all_to_all_single(
+        gathered_input,
+        local_tensor,
+        input_split_sizes=input_split_sizes,
+        output_split_sizes=output_split_sizes,
+        group=pg,
+    )
+    if output_rearrange:
+        rearrage_list = []
+        gathered_list = torch.split(gathered_input, output_split_sizes)
+        for rank in placement_ranks:
+            rearrage_list.append(gathered_list[rank])
+        gathered_input = torch.cat(rearrage_list).contiguous()
+
+    return gathered_input.transpose(0, dim), chunk_local_split
+
+
+def register_chunk_op(op):
+    @sharded_op_impl(op)
+    def sharded_chunk(types, args=(), kwargs=None, pg=None):
+        """
+        Handles ``__torch_function__`` dispatch for the chunk op.
+        If we chunk by the non-sharding dim, we just directly chunk the
+        local tensor and create a list of sharded tensor based on them.
+
+        If we chunk by the sharding dim, we first gather all local shards
+        together, then perform the chunk and shard the tensor for each
+        chunk separately. In reality, we combine all three steps as one
+        communication to save communication cost.
+
+        Args: same as ``torch.chunk``.
+
+        Return:
+            List[ShardedTensor]: Chunk results as a list of ShardedTensor.
+        """
+        st = args[0]
+        chunk_num = args[1]
+        dim = kwargs.get("dim")
+        dim = dim if dim else 0
+
+        # Validate types
+        if not isinstance(st, ShardedTensor):
+            raise TypeError(
+                f"torch function '{op.__name__}', with args: {args} and "
+                f"kwargs: {kwargs} are called for non ShardedTensor!"
+            )
+        spec = st.sharding_spec()
+        if not isinstance(spec, ChunkShardingSpec):
+            raise NotImplementedError("Only ChunkShardingSpec is supported for chunk.")
+        st_size = list(st.size())  # type: ignore[arg-type]
+        sharding_dim = spec.dim
+        local_tensor = st.local_tensor()
+        chunks = torch.arange(st_size[dim]).chunk(chunk_num)
+        chunk_split_size = [len(chunk) for chunk in chunks]
+        local_tensor_chunks = torch.chunk(local_tensor, chunk_num, dim)
+
+        # When performing chunk on the dim of sharding, we need to shuffle the tensor.
```

Comment:
So this is actually more complex than I though it is for the following reasons. Let's take the case of a regular torch.chunk on 1d Tensor of size 100. If we create 4 chunks, then it returns 4 tensors, where tensor 1 is 0-24, tensor 2 is 25-49 and so on. Now the same semantics should hold for ShardedTensor.

Now imagine a ShardedTensor again of size 100 and 1d but sharded across for ranks. So rank 1 has 0-24, rank 2 25-49 and so on.

Now, we if we call torch.chunk on this tensor with num_chunks=4, then it should create 4 ShardedTensors where tensor 1 is 0-24, tensor 2 is 25-49 and so on.

Now this means that ShardedTensor1 only exists on rank0, ShardedTensor2 only on rank 1 and so on. So the sharding spec has actually changed.

I feel for now it is probably best to error out in this case and only allow chunk on the sharded dimension. I'm assuming that is what we need for Transformers right?

Reply:
> Now this means that ShardedTensor1 only exists on rank0, ShardedTensor2 only on rank 1 and so on. So the sharding spec has actually changed.

But this is not the expected behavior right? 

We just first chunk them into 4 parts and do the sharding again. 

For transformers, we can error out it but I guess since we have already done it, I just include it in the PR.

Reply:
Discussed offline, we will make this error out for now.